Cookbook/Proposals/FeatureSwitchesRedesign.md

```diff
@@ -0,0 +1,218 @@
+
+
+## Proposed solution
+
+To address these issues we suggest the following changes in our feature switches architecture:
+
+1. Feature toggles and static configurations related to the particular feature should be defined in this feature module instead of shared module or in the app. This follows interface segregation principle and goes along with improving modularisation - one won't need to change common module with feature switches and recompile all of it (and potentially everything else) when they only need to recompile module that uses this feature switch. This way we also configure our frameworks the same way as we configure any other 3rd party framework and can use the similar infrastrauctore for that (i.e. dedicated instances responsibule for configuring each feature rather than doing that all in the app delegate).
+
+2. Replace `AppConfigurationProtocol` and all related protocols with structs. This will decrease amount of boilerplate we have to write when adding new application configuration. Application configuration will be only concerned with application level configurations which are not directly related to any feature module.
+
+3. All feature switches access points should be moved to the `Current` and consequentially should be only used by builders - view models, flow controllers and renderers should have them injected through their constructors.
+
+4. Root.plist should be generated based on the code that declares feature switches.
+
+## Detailed design
+
+### Module specific feature switches
+
+Each type of feature switch will be implemented with it's own type to make its semantics clear. We will need `LocalFeatureSwitch` and `RemoteFeatureSwitch` types. When later we introduce a new produt configurator service we defined new type `ProductFeatureSwitch`. They all will have a common `value` property and each provider will have a `get(valueForKey: String)` method and should store their properties in a dictionary.
+
+```swift
+// In BabylonDependencies
+
+public struct RemoteFeatureSwitch<T> {
+    let provider = Current.abTestingService
+    var value: T {
+        return provider.get(valueForKey: self.key)
+    }
+    // Name to be used to generate entries for Root.plist using SwiftSyntax
+    let name: String
+    // key to query the value from provider and key for UserDefaults
+    let key: String
+}
+
+public struct LocalFeatureSwitch<T> {
+    let provider = Current.userDefaults
+    var value: T { ... }
+    let name: String
+    let key: String
+}
+
+public struct ProductFeatureSwitch<T> {
+    let provider = Current.productConfigurator
+    var value: T { ... }
+    let name: String
+    let key: String
+}
+```
+
+
+Using dedicated switch types we will be also able to explicitly specify what remote feature switches are A/B tested and which not:
+
+```swift
+typealias ABTest = RemoteFeatureSwitch
```

Comment:
Could you elaborate on this? I'm not sure how it would be defined and based on description I'm thinking that not all `RemoteFeatureSwitch` are A/B tests.

Reply:
yes, but that's true the other way around, all a/b test flags are remote feature switches. 
The alternative is to implement `ABTest` alike `RemoteFeatureSwitch` which will make them not interchangeable which is also fine because implementation is trivial and we are likely to use different services for remote flags and for at tests in future.